server/midddleware/metrics.go

```diff
@@ -25,84 +26,78 @@ import (
 	"google.golang.org/grpc"
 )
 
-func receiveMessage(fullMethod string) {
-	metrics.GetServerRequestCounter(fullMethod, metrics.ServerRequestsReceivedTotal).Counter.Inc(1)
+func increaseCounter(fullMethod string, methodType string, counterName string, value int64) {
+	tags := metrics.GetPreinitializedTagsFromFullMethod(fullMethod, methodType)
+	metrics.Requests.Tagged(tags).Counter(counterName).Inc(value)
 }
 
-func handleMessage(fullMethod string) {
-	metrics.GetServerRequestCounter(fullMethod, metrics.ServerRequestsHandledTotal).Counter.Inc(1)
+func countReceivedMessage(fullMethod string, methodType string) {
+	increaseCounter(fullMethod, methodType, "received", 1)
 }
 
-func errorMessage(fullMethod string) {
-	metrics.GetServerRequestCounter(fullMethod, metrics.ServerRequestsUnknownErrorTotal).Counter.Inc(1)
+func countHandledMessage(fullMethod string, methodType string) {
+	increaseCounter(fullMethod, methodType, "handled", 1)
 }
 
-func specificErrorMessage(fullMethod string, methodType, errSource string, errCode string) {
-	var counter *metrics.ServerRequestCounter
-	if _, ok := metrics.ServerRequestErrorCodeCounters[fullMethod]; !ok {
-		metrics.ServerRequestErrorCodeCounters[fullMethod] = make(map[string]map[string]*metrics.ServerRequestCounter)
-	}
-	if _, ok := metrics.ServerRequestErrorCodeCounters[fullMethod][errSource]; !ok {
-		metrics.ServerRequestErrorCodeCounters[fullMethod][errSource] = make(map[string]*metrics.ServerRequestCounter)
-	}
-	if _, ok := metrics.ServerRequestErrorCodeCounters[fullMethod][errSource][errCode]; !ok {
-		counter = metrics.GetSpecificErrorCounter(fullMethod, methodType, errSource, errCode)
-		metrics.ServerRequestErrorCodeCounters[fullMethod][errSource][errCode] = counter
-	} else {
-		counter = metrics.ServerRequestErrorCodeCounters[fullMethod][errSource][errCode]
-	}
-	counter.Counter.Inc(1)
+func countOkMessage(fullMethod string, methodType string) {
+	increaseCounter(fullMethod, methodType, "ok", 1)
 }
 
-func okMessage(fullMethod string) {
-	metrics.GetServerRequestCounter(fullMethod, metrics.ServerRequestsOkTotal).Counter.Inc(1)
+func countUnknownErrorMessage(fullMethod string, methodType string) {
+	increaseCounter(fullMethod, methodType, "unknown_error", 1)
 }
 
-func getTimeHistogram(fullMethod string) *metrics.ServerRequestHistogram {
-	return metrics.GetServerRequestHistogram(fullMethod, metrics.ServerRequestsHandlingTimeBucket)
+func countSpecificErrorMessage(fullMethod string, methodType, errSource string, errCode string) {
+	metaData := metrics.GetGrpcEndPointMetadataFromFullMethod(fullMethod, methodType)
+	tags := metaData.GetSpecificErrorTags(errSource)
```

Comment:
if we create subscope for the `errSource` then we don't need it?

Reply:
I did this.